Feat: oidc-integration

er-santosh opened the pull request:
Tasks done
---

- Implemented Oidc Client for oidc integration.
- Added OIDC disable feature.
- Added Oidc session middleware.

Issue: [https://github.com/bigcapitalhq/bigcapital/issues/237] (Support for OpenID Connect (OIDC) / OAuth2)

er-santosh committed:
feat: oidc-integration

Tasks done
---

- Implemented Oidc Client for oidc integration.
- Added OIDC disable feature.
- Added oidc session middleware.

er-santosh committed:
revert: reverted sorted dependencies from server package json

poudelprakash commented on packages/webapp/src/lang/ar/index.json:
```diff
@@ -1591,8 +1592,8 @@
-  "landed_cost.dialog.label_unallocated_cost_amount":"قيمة التكلفة غير المحملة:",
-  "landed_cost.error.the_total_located_cost_is_bigger_than_the_transaction_line":"إجمالي قيمة التكلفة المحملة أكبر من قيمة سطر المعاملة.",
+  "landed_cost.dialog.label_unallocated_cost_amount": "قيمة التكلفة غير المحملة:",
+  "landed_cost.error.the_total_located_cost_is_bigger_than_the_transaction_line": "إجمالي قيمة التكلفة المحملة أكبر من قيمة سطر المعاملة.",
```
lets revert unintended changes in this file

er-santosh commented on packages/webapp/src/hooks/query/authentication.tsx:
```diff
@@ -37,6 +38,64 @@ export const useAuthLogin = (props) => {
+      // Set authentication cookies.
+      setAuthLoginCookies(data.data);
+
+      window.location.href = '/';
```
@abouolia Previously while logging in, the login page was reloaded after the successful login. Here in oidc login we are supposed to get code as query param in login url so we are redirecting it to '/' which is a bit static for now as reloading the route will make api call for authorization callback. Do we have any standard path that needs to be added here after login or what will be the best use case here?

er-santosh commented on packages/webapp/src/containers/Setup/SetupLeftSection.tsx:
```diff
@@ -38,11 +38,11 @@ function SetupLeftSectionFooter() {
 
   // Handle logout link click.
   const onClickLogout = () => {
-    setLogout();
```
@abouolia I have created custom logout hook for oidc(useAuthOidcLogout) where setLogout will be called after the oidc logout api call in both condition (error or success). Is it ok to follow this?

er-santosh commented on packages/webapp/src/hooks/query/authentication.tsx:
```diff
@@ -37,6 +38,64 @@ export const useAuthLogin = (props) => {
+ * Authentication OIDC logout
+ */
+
+export const useAuthOidcLogout = (props) => {
+  const apiRequest = useApiRequest();
+  const { setLogout } = useAuthActions();
+
+  return useMutation((values) => apiRequest.post('oidc/logout', values), {
+    select: (res) => res.data,
+    onSuccess: (data) => {
+      const logoutUrl = data.data.logout_url;
+
+      setLogout(logoutUrl);
+    },
+    onError: (err) => {
+      setLogout();
+    },
```
@abouolia Here I have implemented oidc logout such that after successful oidc login, the session end url will be passed to setLogout and redirected in same way else the normal logout will work accordingly. Is the best way to be followed?

er-santosh commented on packages/webapp/src/hooks/state/authentication.tsx:
```diff
@@ -32,7 +32,11 @@ export const useAuthActions = () => {
       removeAuthenticationCookies();
 
-      window.location.reload();
+      if (href) {
+        window.location.href = href;
+      } else {
+        window.location.reload();
+      }
```
@abouolia Here the only change is, if the session_end url is passed, it will be redirected to that url and gets callback from OIDC provider to post logout uri else just reload the page.

er-santosh commented on packages/server/src/services/Authentication/_utils.ts:
```diff
@@ -2,12 +2,18 @@ import JWT from 'jsonwebtoken';
 
+interface IExtendedSystemUser extends ISystemUser {
+  oidc_access_token?: string;
+  oidc_id_token?: string;
+  oidc_refresh_token?: string;
+}
+
 /**
  * Generates JWT token for the given user.
- * @param {ISystemUser} user
+ * @param {IExtendedSystemUser} user
  * @return {string} token
  */
-export const generateToken = (user: ISystemUser): string => {
+export const generateToken = (user: IExtendedSystemUser): string => {
   const today = new Date();
   const exp = new Date(today);
   exp.setDate(today.getDate() + 60);
```
@abouolia I have created new IExtendedSystemUser interface extending the current ISystemUser to store the oidc tokens which will be attached in request token so that it can be used to logout the oidc user and revoke the access tokens. Also it will be used in oidcSession middleware to check if user exist for that oidc access token.

Is it ok to store the oidc tokens in SystemUser table?
What will be the best way to work with the oidc tokens provided by the providers?

er-santosh commented on packages/server/src/loaders/eventEmitter.ts:
```diff
@@ -89,7 +89,7 @@ import { InvoiceChangeStatusOnMailSentSubscriber } from '@/services/Sales/Invoic
 import { SaleEstimateMarkApprovedOnMailSent } from '@/services/Sales/Estimates/subscribers/SaleEstimateMarkApprovedOnMailSent';
 import { DeleteCashflowTransactionOnUncategorize } from '@/services/Cashflow/subscribers/DeleteCashflowTransactionOnUncategorize';
-import { PreventDeleteTransactionOnDelete } from '@/services/Cashflow/subscribers/PreventDeleteTransactionsOnDelete'; }
+import { PreventDeleteTransactionOnDelete } from '@/services/Cashflow/subscribers/PreventDeleteTransactionsOnDelete'; 
```
@abouolia I couldnot build server due to this syntax error here. So i removed it to get it working.

er-santosh committed:
changes: added OIDC environment in production docker compose

er-santosh committed:
feat: added introspection for oidc access token

er-santosh committed:
revert: reverted ar lang indentation

er-santosh committed:
feat: verified oidc access token in jwt auth middleware

er-santosh committed:
feat: show loading overlay when user logs out

er-santosh committed:
feat/get-tenant-of-authenticated-user

Tasks done
---

- Created api endpoint to fetch authenticated user details and its tenant.
- Added cors middleware to allow resources to whitelisted domains.

er-santosh committed:
Merge branch 'develop' into feat/oidc-integration

er-santosh committed:
Merge branch 'feat/oidc-integration' into feat/get-tenant-of-authenticated-user

er-santosh committed:
Merge branch 'develop' into feat/get-tenant-of-authenticated-user

er-santosh committed:
fixed issue after merge

er-santosh committed:
Merge pull request #1 from er-santosh/feat/get-tenant-of-authenticated-user

Feat/get tenant of authenticated user

er-santosh committed:
Merge branch 'develop' into feat/oidc-integration
